[Clean up] Bank accounts/tranasactions

Civolilah opened the pull request:
@turbo124 Just one small thing to mention, modal for new bank account i too big because we did not specified `"regular"` size earlier. It's specified on Add Company branch, so once we merge it, it will look perfect.

beganovich commented on src/pages/settings/bank-accounts/create/Create.tsx:
```diff
@@ -54,17 +57,21 @@ export function Create() {
   const handleCancel = () => {
     if (!isFormBusy) {
-      navigate(route('/settings/bank_accounts'));
+      navigate(route('/settings/bank_accounts/'));
```
As previously mentioned, no need to use `route` when there are no query or URL parameters. It does nothing.

Civolilah replied:
> As previously mentioned, no need to use `route` when there are no query or URL parameters. It does nothing.

Removed

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -196,6 +196,10 @@ export function Create() {
               errorMessage={errors?.bank_integration_id}
+              actionLabel={t('new_bank_account')}
+              onActionClick={() =>
+                navigate(route('/settings/bank_accounts/create'))
```
Drop `route` helper.

Civolilah replied:
> Drop `route` helper.

removed

beganovich commented on src/pages/transactions/edit/Edit.tsx:
```diff
@@ -223,6 +223,10 @@ export function Edit() {
               errorMessage={errors?.bank_integration_id}
+              actionLabel={t('new_bank_account')}
+              onActionClick={() =>
+                navigate(route('/settings/bank_accounts/create'))
```
Drop `route` helper.

Civolilah replied:
> Drop `route` helper.

removed

beganovich reviewed (changes requested):
Looks good, small changes needed :+1:

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
 
```
`event` shouldn't be undefined/nullable here.

Civolilah replied:
removed

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
+      try {
+        await request('POST', endpoint('/api/v1/bank_transactions'), {
+          ...transaction,
+          amount: Number(transaction?.amount),
```
Please remove optional chain. If `transaction` is nullable here, we have much bigger problems.

Civolilah replied:
No, here it's not needed. Removed.

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
+          ...transaction,
+          amount: Number(transaction?.amount),
+          base_type:
+            transaction?.base_type === TransactionType.Deposit
```
Nullable reference on `transaction`.

Civolilah replied:
removed

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
+              ? ApiTransactionType.Credit
+              : ApiTransactionType.Debit,
+        });
+        toast.success(t('created_transaction'));
```
`toast` accepts translatable string. No need to pass it using `t`.

Civolilah replied:
removed

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
+        console.log(axiosError);
+        console.error(axiosError);
+
+        if (axiosError?.response?.status === 422) {
```
Nullable reference.

Civolilah replied:
Same replay for axios error as above.

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
+          setErrors(axiosError?.response?.data?.errors);
+          toast.dismiss();
+        } else {
+          toast.error(t('error_title'));
```
toast is translatable. Just pass plain string, please:

```ts
toast.error('error_title')
```

or in this case:

```ts
toast.error()
```

Civolilah replied:
changed

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -85,33 +90,43 @@ export function Create() {
+        setIsSaving(false);
       }
+    } else {
+      setErrors((prevState) => ({
+        ...prevState,
+        bank_integration_id: t('required_field'),
+      }));
```
I don't think we ever do frontend validation. Errors should come from API.

Civolilah replied:
API in this case returned me 500 error without message instead of 422 with error message so I have to set it manually.

![image](https://user-images.githubusercontent.com/51542191/204322094-aa8007b1-a968-40c5-87a2-5b193e68538f.png)

beganovich commented on src/pages/transactions/edit/Edit.tsx:
```diff
@@ -115,34 +121,47 @@ export function Edit() {
+        setIsSaving(false);
       }
+    } else {
+      setErrors((prevState) => ({
+        ...prevState,
+        bank_integration_id: t('required_field'),
+      }));
     }
```
Please use API validation, not frontend.

Civolilah replied:
replied above

beganovich commented on src/pages/transactions/edit/Edit.tsx:
```diff
@@ -115,34 +121,47 @@ export function Edit() {
+
+        console.error(axiosError);
+
+        if (axiosError?.response?.status === 422) {
```
Nullable reference.

Civolilah replied:
replied above

beganovich commented on src/pages/transactions/edit/Edit.tsx:
```diff
@@ -115,34 +121,47 @@ export function Edit() {
-        toast.dismiss();
-      } else {
-        toast.error(t('error_title'));
+    if (transaction?.bank_integration_id) {
```
Please ensure `transaction` is not nullable before accessing this block of code.

Civolilah replied:
removed, but I had to add default object to make sure that transaction will never be undefined.

beganovich commented on src/pages/transactions/edit/Edit.tsx:
```diff
@@ -115,34 +121,47 @@ export function Edit() {
+          endpoint('/api/v1/bank_transactions/:id', { id }),
+          {
+            ...transaction,
+            amount: Number(transaction?.amount),
+            base_type:
+              transaction?.base_type === TransactionType.Deposit
```
Nullable reference.

beganovich commented on src/pages/transactions/edit/Edit.tsx:
```diff
@@ -115,34 +121,47 @@ export function Edit() {
+          setErrors(axiosError?.response?.data?.errors);
+          toast.dismiss();
+        } else {
+          toast.error(t('error_title'));
```
`toast.error()` is enough here.

Civolilah replied:
changed

Civolilah commented on src/pages/settings/bank-accounts/create/Create.tsx:
```diff
@@ -56,15 +58,15 @@ export function Create() {
   const handleCancel = () => {
     if (!isFormBusy) {
-      navigate(route('/settings/bank_accounts'));
+      navigate('/settings/bank_accounts/');
```
removed

Civolilah commented on src/pages/transactions/common/validation/ValidationInterface.tsx:
```diff
@@ -9,10 +9,10 @@
-  currency_id: string;
-  bank_integration_id: string;
-  description: string;
+  base_type?: string;
+  date?: string;
+  amount?: string;
+  currency_id?: string;
+  bank_integration_id?: string;
+  description?: string;
```
The reason for that is this block of code, 

```
setErrors((prevState) => ({
        ...prevState,
        bank_integration_id: t('required_field'),
 }));
```

If I don't keep them optional, I will have to enter all of them but I only want to enter bank_integration_id. The reason I set up manual validation on the frontend is because the API returns a 500 error without any validation message instead of a 422 and with a validation message for this field, so I had to do it manually.

Civolilah commented on src/pages/transactions/components/CreateBankAccountModal.tsx:
```diff
@@ -0,0 +1,136 @@
+
+  const handleSave = async (event: FormEvent<HTMLFormElement>) => {
+    if (!isFormBusy) {
+      event?.preventDefault();
```
removed

Civolilah commented on src/pages/transactions/components/CreateBankAccountModal.tsx:
```diff
@@ -0,0 +1,136 @@
+        console.error(error);
+
+        if (error?.response?.status === 422) {
+          setErrors(error?.response?.data?.errors);
```
> Remove optional chain where not needed.

I replied above.

turbo124 commented:
@Civolilah 

if you see a 500, that needs to be raised as the priority issue, we should not work around the 500 in the UI.

Civolilah commented:
> @Civolilah
> 
> if you see a 500, that needs to be raised as the priority issue, we should not work around the 500 in the UI.

@turbo124 I've tested the request multiple times and currently if I don't enter the `bank_integration_id` the API will return a 500 error. So to do the validation here I had to set it manually. Maybe it's better to look at error handling on the API side and pass the error for `bank_integration_id` as another (422) to make it easier to display and be consistent?

beganovich commented:
Is this ready for review? Does it need additional change related to the 500? @Civolilah @turbo124

Civolilah commented:
> @beganovich I've fixed the 500 issue from my side.

@turbo124 @beganovich  I just updated the PR with all the custom validation stuff. Also, I checked several times the error we get from the API and with several variants of the `BankTransaction` object. But I seem to get a 500 error if I don't pass the `bank_integration_id` property to the object and if I set it to `bank_integration_id: ''`. In both cases we get 500 errors. Here's what it looks like:

![image](https://user-images.githubusercontent.com/51542191/204631193-701bcabb-d9fb-4838-a486-21cb4b8c2ab7.png)

- Also, while working on this task, I tested a lot of bank accounts and bank transaction operations and found that the API does not return a 422 message when editing a bank account. So if we leave `bank_account_name` blank it will work.

Civolilah commented:
@beganovich Through chatting on Slack with @turbo124 we resolved it, and I can confirm that it's tested and validation for bank account now works properly!

beganovich commented on src/pages/settings/bank-accounts/create/Create.tsx:
```diff
@@ -56,38 +56,36 @@ export function Create() {
+          navigate('/settings/bank_accounts');
+        })
+        .catch((error: AxiosError) => {
+          if (error?.response?.status === 422) {
```
I am sorry we have to go thru this again, but `error` is not nullable/undefined property here.

Civolilah replied:
> I am sorry we have to go thru this again, but `error` is not nullable/undefined property here.

Sorry I just realized the response is just null, I tried removing the optional chaining from all objects and it didn't work. Removed!

beganovich commented on src/pages/settings/bank-accounts/edit/Edit.tsx:
```diff
@@ -70,38 +74,38 @@ export function Edit() {
+          navigate('/settings/bank_accounts');
+        })
+        .catch((error: AxiosError) => {
+          if (error?.response?.status === 422) {
```
Please remove nullable reference to `error`.

Civolilah replied:
removed

beganovich commented on src/pages/transactions/components/CreateBankAccountModal.tsx:
```diff
@@ -0,0 +1,127 @@
+        .catch((error: AxiosError) => {
+          console.error(error);
+
+          if (error?.response?.status === 422) {
```
Nullable reference to `error`.

Civolilah replied:
removed

beganovich commented on src/pages/transactions/create/Create.tsx:
```diff
@@ -81,38 +92,34 @@ export function Create() {
+        navigate('/transactions');
+      })
+      .catch((error: AxiosError) => {
+        if (error?.response?.status === 422) {
```
Nullable refrence to `error.`

Civolilah replied:
removed

beganovich reviewed (changes requested):
Looks good, small changes needed.

Civolilah commented:
> Additional changes:
> 
> Modal size:
> 
> * Can we please use regular modal size? There's no reason for modal to be this big.
> * Change `bank_account_name` to `name`. We are already in context of bank, so reference is obvious.
> * Cancel button has this weird background. I suggest removing it as we have cancel on the top right and modal can be self-closed by clicking away.
> 
> Reuse create logic:
> 
> * Can we extract and reuse request for creating bank accounts? I can see exact code duplicated in two components.

@beganovich Resolved all suggestions! But for size of modal, this is how it looks at my end:

![image](https://user-images.githubusercontent.com/51542191/204936413-5db05052-5ae3-4b16-92a7-d96b4f266d48.png)

It looks like a perfect size on my end, do you think this size is large or did you have the larger version?

beganovich commented:
Even on your screenshot that looks unnecessarily big. 

Here's the updated version (regular size).

![image](https://user-images.githubusercontent.com/13711415/205053386-6679b66b-59a0-41e0-bcc2-4d743eeab363.png)
